test/azul_test_case.py

```diff
@@ -1,10 +1,49 @@
-import os
-from unittest import TestCase, mock
+from unittest import TestCase
 
 import boto3.session
 
 
-class AzulTestCase(TestCase):
+class SafeTestCase(TestCase):
+    """
+    SafeTestCase makes sure that tearDown / cleanup methods are always run when
+    They should be.
+
+    This means that
+    - if a KeyboardInterrupt is raised in a test, then tearDown, tearDownClass,
+      and tearDownModule will all still run
+    - if any exception is raised in a setUp, then tearDown, tearDownClass,
+      and tearDownModule will all still run
+
+    Caveats:
+    - All tearDown methods MUST be idempotent. It's possible a step of tearDown
+      will run without it's corresponding setUp having run.
+    - If an exception is raised in SetUpClass or SetUpModule, tearDowns will
+      not be run. (Working around this involves making a custom TestSuite which
+      will be ignored if you run tests through PyCharm)
+    """
+
+    def run(self, result=None):
+        test_method = getattr(self, self._testMethodName)
+        wrapped_test = self._cleanup_wrapper(test_method, KeyboardInterrupt)
+        setattr(self, self._testMethodName, wrapped_test)
+
+        self.setUp = self._cleanup_wrapper(self.setUp, BaseException)
```

Comment:
What about exceptions in setupClass()? Should an exception in that method still lead to a call of tearDownClass()?

Reply:
@hannes-ucsc see the second caveat on line 20.